Change all the things.

elijahsummers opened the pull request:
- Remove old libraries/config.
- Refactor Manifests.
- Remove code related to building assets from the helper.
- Allow hierarchical manifests.
- Allow blessing (splitting) css.
- Update options.
- Cleanup readme.
- Improve handling of assets with missing extensions.
- Use exceptions on error (rather than exit).
- Allow checking less imports for changes (sloooooow).

Yea... It's a big one.

Commit:
Change all the things.

* Remove old libraries/config.
* Refactor Manifests.
* Remove code related to building assets from the helper.
* Allow hierarchical manifests.
* Allow blessing (splitting) css.
* Update options.
* Cleanup readme.
* Improve handling of assets with missing extensions.
* Use exceptions on error (rather than exit).
* Allow checking less imports for changes (sloooooow).

elijahsummers commented on Manifest.php:
```diff
@@ -48,117 +38,294 @@ public static function all() {
+		}
+		$assets = $config['manifests'][$type][$name];
+		return new Manifest($type, $name, $assets, $options);
+	}
```
Added this to mimic the functionality of the old constructor (create a Manifest based on a manifest name), since I changed the constructor to accept a list of assets.

elijahsummers commented on Manifest.php:
```diff
@@ -48,117 +38,294 @@ public static function all() {
+					if (!isset($seenAssets[$item])) $seenAssets[$item] = true;
+				}
+			}
+			$this->filenames = array_keys($seenAssets);
```
If any of the assets in a manifest are of the form `[name].manifest`, and there is a manifest named `[name]`, then load assets from that manifest. Has some very basic cycle protection via `$seenManifests`.

elijahsummers commented on Manifest.php:
```diff
@@ -48,117 +38,294 @@ public static function all() {
-		$this->config = Libraries::get('li3_frontender');
-		if(!isset($this->config['manifests'][$type][$name])) {
-			throw new Exception("manifest $type/$name not found");
+		if ($this->requiresFullAssetNames()) {
```
Only perform this early, intensive, stuff if the options don't look "production-ish".

elijahsummers commented on Manifest.php:
```diff
@@ -48,117 +38,294 @@ public static function all() {
+			throw new Exception("Failed checking less imports for `{$fullPath}`.");
+		}
+		return $result;
+	}
```
Finding a ton of less dependencies is understandably slow (especially on the devup VMs). Would love to move as many of the `@import`s as possible into the manifest configs.

elijahsummers commented on Manifest.php:
```diff
@@ -205,28 +363,56 @@ protected function compileFile($path, $destination) {
+			$files[] = $out;
+		}
+		return $files;
+	}
```
`blessc` doesn't do a wonderful job of naming split files. This is here as a workaround to ensure consistency, i.e., that every processed asset filename begins with its original filename (including ext).

elijahsummers commented on config/bootstrap.php:
```diff
@@ -1,85 +0,0 @@
- */
-require __DIR__ . '/libraries.php';
-
-Dispatcher::applyFilter('run', function($self, $params, $chain) use ($options) {
```
I don't see how this filter could possibly be used (at least in any recent version of frontender). Requests for webroot assets that exist don't even touch lithium...

elijahsummers commented on extensions/helper/Assets.php:
```diff
@@ -43,71 +28,23 @@ public function style($manifest) {
-		$manifest->verbose = $this->debug;
-
-		if($this->config['batch_only']) {
-			$files = array($this->mangledFilename());
```
This right here scares the crap out of me. By my reasoning `$this->mangledFilename()` should throw an error in production. So how were these helpers ever working?!?!

elijahsummers commented:
Also, this will resolve #2

elijahsummers commented:
If you want to test locally: checkout this branch in `app/libraries/li3_frontender`, then fiddle with these [options](https://github.com/mdx-dev/li3_frontender/blob/frontender-bender/README.md#configuration-options) (see https://github.com/mdx-dev/vitals/compare/feature;test-frontender for reference)

bayleedev commented:
:+1: I'm excited about the new changes!

nwp commented:
Is there any way to unit test this?

bayleedev commented:
Agreed, I'd love to see some unit tests here. Maybe even some integration tests using [vfsStream](https://github.com/mikey179/vfsStream)? :D

Lemme know if you'd like to pair on this. We can get the unit tests started and setup on travis.

elijahsummers commented:
Absolutely. @blainesch after standup, would you be able to walk me through setting up travis for an OSS repo a la charizard?

Commit:
Setup travis file and basic unit test.

Commit:
Merge pull request #5 from mdx-dev/feature/tests

Setup frontender on travis.

bayleedev commented:
@elijahsummers I've found an issue with the new updates.

When I get the scripts for a manifest:

```
<?= $this->assets->script('global') ?>
```

The first line [here](https://github.com/mdx-dev/li3_frontender/blob/frontender-bender/extensions/helper/Assets.php#L39-L40) returns the correct manifest, but files isn't in the right order.

Commit:
Try out vfs for testing.

Commit:
Add test tests and run tests from the app.

Commit:
Fixes for asset ordering.

- Assets are no longer reversed.
- Fixed an issue where lazy asset names in embedded manifests could cause
  duplicates in the final list of assets.
- Added tests.

elijahsummers commented:
_bump_ How's it looking?

bayleedev commented on tests/cases/extensions/command/AssetTest.php:
```diff
@@ -0,0 +1,13 @@
+class AssetTest extends \lithium\test\Unit {
+
+	public function testTrue() {
+		$this->assertTrue(true);
```
You can probably delete this test [:

elijahsummers replied:
all right...but I won't be held responsible the day `true` no longer is true!!

bayleedev replied:
I'll take the fall for that one [:

elijahsummers committed:
Remove sample test.

dolinsky commented:
![](http://imgs.xkcd.com/comics/idiocracy.png)

bayleedev commented on tests/cases/ManifestTest.php:
```diff
@@ -0,0 +1,149 @@
+				),
+				'embeddedOrdered' => array(
+					'begin',
+					'ordered.manifest',
```
That's badass.

elijahsummers replied:
ain't it? I'm hoping it makes organizing assets much easier.

bayleedev commented:
:+1: Thanks for the hard work!

brandonwestcott commented on Manifest.php:
```diff
@@ -48,117 +38,290 @@ public static function all() {
 
-		$this->less_binary = 'lessc';
-		$this->coffee_binary = "coffee";
+	public function build() {
```
This is the ruby styles coming out, but....

Is there any way to break this method down into smaller functional pieces in the same class or separate into a helper class with smaller class methods? Even if not for reuse, I believe it helps bring clarity to what "build" actually does

bayleedev replied:
I agree this should be broken up, since this is a large method. However there are lots of large methods which all cannot be done, and this pull request has already done quite a bit of cleanup including adding a good handful of unit tests.

brandonwestcott commented on tests/cases/ManifestTest.php:
```diff
@@ -0,0 +1,149 @@
+<?php
```
Love it!

Commit:
Split out the larger segments of build().
